packages/core/src/utils/make-namespaced.ts

```diff
@@ -44,52 +44,74 @@ export function makeNamespacedInput(input: SolcInput, output: SolcOutput): SolcI
-
-              modifications.push(makeInsertAfter(contractNode, insertText));
+          const contractNodes = contractDef.nodes;
+          for (const contractNode of contractNodes) {
+            switch (contractNode.nodeType) {
+              case 'ErrorDefinition':
+              case 'EventDefinition':
+              case 'FunctionDefinition':
+              case 'ModifierDefinition':
+              case 'VariableDeclaration': {
+                if (contractNode.documentation) {
+                  modifications.push(makeDelete(contractNode.documentation, orig));
+                }
+                modifications.push(makeDelete(contractNode, orig));
+                break;
+              }
+              case 'UsingForDirective': {
+                modifications.push(makeDelete(contractNode, orig));
+                break;
+              }
+              case 'StructDefinition': {
+                const storageLocation = getStorageLocationAnnotation(contractNode);
+                if (storageLocation !== undefined) {
+                  const structName = contractNode.name;
+                  const variableName = `$${structName}_${(Math.random() * 1e6).toFixed(0)}`;
+                  const insertText = ` ${structName} ${variableName};`;
+
+                  modifications.push(makeInsertAfter(contractNode, insertText));
+                }
+                break;
+              }
+              case 'EnumDefinition':
+              case 'UserDefinedValueTypeDefinition':
+              default: {
+                // - EnumDefinition may be used in structures with storage locations
+                // - UserDefinedValueTypeDefinition may be used in structures with storage locations
+                // - default: in case unexpected ast nodes show up
+                break;
+              }
             }
           }
+          break;
         }
-      } else if (isNodeType('FunctionDefinition', node) || isNodeType('VariableDeclaration', node)) {
-        if (node.documentation) {
-          modifications.push(makeDelete(node.documentation, orig));
+        case 'EnumDefinition':
+        case 'ImportDirective':
+        case 'PragmaDirective':
+        case 'StructDefinition':
+        case 'UserDefinedValueTypeDefinition':
+        default: {
+          // - EnumDefinition may be used in structures with storage locations
+          // - ImportDirective may import types used in structures with storage locations
+          // - PragmaDirective is necessary for compilation
+          // - StructDefinition may be used in structures with storage locations
+          // - UserDefinedValueTypeDefinition may be used in structures with storage locations
+          // - default: in case unexpected ast nodes show up (file-level events since 0.8.22)
+          break;
         }
-        // Replace with a dummy variable of arbitrary type
-        const name = node.name;
-        const insertText = `uint256 constant ${name} = 0;`;
```

Comment:
This was replaced by `uint256 constant` because the function definition or file level constant could be imported by another file, and we are not removing imports from the other file. I've added a testcase to reflect this.

Therefore we cannot outright delete it.  We cannot also just keep the original function or constant's source code, because they might reference deleted functions.

What we could consider is replace with a dummy variable but avoid duplicates if one already exists with the same name.

Reply:
That is a great point. I forgot about explicit imports :/